AddWebResourceRequestedFilter for workers (#2499)
MicrosoftEdge/WebView2Feedback: specs/WebResourceRequestedWorkers.md

> WebResourceRequested events for workers

Reviewer: If there's an existing API doc for WebResourceRequested already checked in, and you'd like to update that doc rather than create a new doc, that is also an option, but not required.

===

# Background
Currently [WebResourceRequested events](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#add_webresourcerequested) are raised only for main page HTTP requests. We are extending [AddWebResourceRequestedFilter](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#addwebresourcerequestedfilter) API to allow you additionally to subscribe to events raised from Service or Shared workers and fixing oop iframes events as part of the main page.

# Examples
Subscribe to WebResourceRequested event for service workers

.Net
```c#
Controller.CoreWebView2.AddWebResourceRequestedFilter("*", CoreWebView2WebResourceContext.All, CoreWebView2WebResourceRequestSource.ServiceWorker);
Controller.CoreWebView2.WebResourceRequested += (sender, e) =>
{
    string message =  "{" + WebResourceRequestedToJsonString(e.Request, e.RequestSource);

>     message += WebViewPropertiesToJsonString(sender) + "}";

Reviewer: This looks like sort of logging. Do we have an example of a scenario an end developer wants to do with WebResourceRequested for service workers that we can demonstrate in the sample code instead?

Author: Updated sample with overriding image in request

                message += WebViewPropertiesToJsonString(m_webviewEventSource.get());
                message += L"}";
                PostEventMessage(message);
                return S_OK;
            })
            .Get(),
        &m_webResourceRequestedToken);
```

# API Details
## IDL
```c#
/// Specifies the source of `WebResourceRequested` event.
[v1_enum]
typedef enum COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE {
  /// Indicates that web resource is requested from main page including dedicated workers and iframes.
  COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_MAIN = 1,

  /// Indicates that web resource is requested from shared worker.
  COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_SHARED_WORKER = 2,

  /// Indicates that web resource is requested from service worker.
  COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_SERVICE_WORKER = 4,

  /// Indicates that web resource is requested from any supported source.
  COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_ALL = 0XFFFFFFFF
} COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE;
cpp_quote("DEFINE_ENUM_FLAG_OPERATORS(COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE);")

[uuid(1cc6a402-3724-4e4a-9099-8cf60f2f93a1), object, pointer_default(unique)]

> interface ICoreWebView2: IUnknown {

Reviewer: This should be something like
```suggestion
interface ICoreWebView2_14: ICoreWebView2_13 {
```

>   /// Extends AddWebResourceRequestedFilter to support events from shared

Reviewer: This will become the reference documentation so we don't want to phrase this like its an update to a previous method. Plenty of devs will read this for the first time together with the other documentation.  You can copy and edit the existing AddWebResourceRequestedFilter documentation if you like.

>   /// and service workers. To receive all raised events filters have to be added

Reviewer: > To receive all raised events filters have to be added

Is this comment about workers or is it also applicable to AddWebResourceRequestedFilter? It makes sense that you would need to set filters before you navigate in order to get the web resource requests for that navigate.

Author: So from implementation POV for main page we can subscribe anytime and event should be received - but we are not stating this anywhere in documentation. For workers and oopifs(now part of main page event) we need to subscribe before navigation.

  /// before main page navigation.

>   /// Because service worker runs separately from main page its WebResourceRequested

Reviewer: ```suggestion
  /// Because service workers and shared workers run separately from any one HTML document WebResourceRequested
```

Author: will fix, but this is applicable only for service workers, not shared

>   /// will be raised for all webviews which are subscribed to

Reviewer: ```suggestion
  /// will be raised for all CoreWebView2s that have appropriate filters added in the corresponding CoreWebView2Environment.
```

  /// COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_SERVICE_WORKER request source.

>   /// It is recommended to enable WebResourceRequested filter for service workers

Reviewer: ```suggestion
  /// You should only add a WebResourceRequested filter for COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_SERVICE_WORKER or COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_SHARED_WORKER 
```

>   /// only on a main WebView to avoid handling the same WebResourceRequested

Reviewer: ```suggestion
  /// on one CoreWebView2 to avoid handling the same WebResourceRequested
```

  /// event multiple times.
  ///
  /// For more information about web resource requested filters, navigate to
  /// AddWebResourceRequestedFilter.
  HRESULT AddWebResourceRequestedFilterWithRequestSource(
    [in] LPCWSTR const uri,
    [in] COREWEBVIEW2_WEB_RESOURCE_CONTEXT const resourceContext,

>     [in] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE const requestSource);

Reviewer: What happens if requestSource is 0? We fail?

Author: We will not fail - just ignore it(will fix one small issue). How should we properly handle it?

Reviewer: Such a filter will never match and is almost certainly an error on part of the end developer. Would be good to fail in that case (and document it)

  /// Removes a matching WebResource filter that was previously added for the
  /// `WebResourceRequested` event.  If the same filter was added multiple
  /// times, then it must be removed as many times as it was added for the
  /// removal to be effective.  Returns `E_INVALIDARG` for a filter that was
  /// never added.
  HRESULT RemoveWebResourceRequestedFilterWithRequestSource(
      [in] LPCWSTR const uri,
      [in] COREWEBVIEW2_WEB_RESOURCE_CONTEXT const resourceContext,
      [in] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE const requestSource);
 }

/// Event args for the `WebResourceRequested` event.
[uuid(572f38f9-c317-4b6c-8dd2-c8b894779bbb), object, pointer_default(unique)]
interface ICoreWebView2WebResourceRequestedEventArgs : IUnknown {
  /// The web resource request source.
  [propget] HRESULT RequestSource(
  [out, retval] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE* requestSource);
}
```

## .NET and WinRT
```c# (but really MIDL3)
namespace Microsoft.Web.WebView2.Core
{
    [flags]
    enum CoreWebView2WebResourceRequestSource
    {

>         Main = 0x00000001,

Reviewer: The word 'main' is sort of ambiguous here. Does this really mean like 'Document'? Or it also includes child frames so maybe 'Documents'?
```suggestion
        Documents = 0x00000001,
```

Author: Documents are in plural but worker(SharedWorker...)  is in singular - is it ok? Or make workers plural also?

Author: ` /// Indicates that web resource is requested from main page including dedicated workers and iframes.`

        SharedWorker = 0x00000002,
        ServiceWorker = 0x00000004,
        All = 0xFFFFFFFF,
    };

    runtimeclass CoreWebView2
    {

>         void AddWebResourceRequestedFilter(

Reviewer: If you copy this out of the generated Microsoft.Web.WebView2.Core.IDL there should be a block with an interface name containing these two new methods? We should include that here to make it clear we aren't breaking compat

Author: Just wanted to remind you that implementation for new method will fix OOP iframes events for default value, for which WebResourceRequested is not sent in old method.  Let me know if we want to keep the old name `AddWebResourceRequestedFilter`, or have new API `AddWebResourceRequestedFilterWithRequestSource`.

Reviewer: No complaint about the name. Just want to see something like the following around the new methods which you should see in the generated IDL

```
[interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2_16")]
{
...
}
```

                String uri,
                CoreWebView2WebResourceContext ResourceContext,

>                 CoreWebView2WebResourceRequestSource ResourceSource);

Reviewer: This should be either ResourceRequestSource or RequestSource right? (Also in a few other places)
```suggestion
                CoreWebView2WebResourceRequestSource RequestSource);
```

                        CHECK_FAILURE(m_webviewEventSource->QueryInterface(IID_PPV_ARGS(&webview2)));
                        CHECK_FAILURE(webview2->get_Environment(&environment));
                        CHECK_FAILURE(SHCreateStreamOnFileEx(
                            file_path, STGM_READ, FILE_ATTRIBUTE_NORMAL,
                            FALSE, nullptr, &response_stream));
                        Microsoft::WRL::ComPtr<ICoreWebView2WebResourceResponse> response;
                        CHECK_FAILURE(environment->CreateWebResourceResponse(
                            response_stream.Get(), 200, L"OK", L"Content-Type: image/jpeg",
                            &response));
                        args->put_Response(response.Get());
                    }
                }
                return S_OK;
            })
            .Get(),
        &m_webResourceRequestedToken);
```

# API Details
## IDL
```c#
/// Specifies the source of `WebResourceRequested` event.
[v1_enum]
typedef enum COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE {
  /// Indicates that web resource is requested from main page including dedicated workers and iframes.

>   COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_DOCUMENTS = 1,

Reviewer: Seeing it here next to the others, this should probably be non-plural

```suggestion
  COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_DOCUMENT = 1,
```

  /// wildcard string for corresponding request source will be raised via
  /// the `WebResourceRequested` event. To receive all raised events filters have
  /// to be added before main page navigation.
  ///
  /// The `uri` parameter value is a wildcard string matched against the URI
  /// of the web resource request. This is a glob style
  /// wildcard string in which a `*` matches zero or more characters and a `?`
  /// matches exactly one character.
  /// These wildcard characters can be escaped using a backslash just before
  /// the wildcard character in order to represent the literal `*` or `?`.
  ///
  /// The matching occurs over the URI as a whole string and not limiting
  /// wildcard matches to particular parts of the URI.
  /// The wildcard filter is compared to the URI after the URI has been
  /// normalized, any URI fragment has been removed, and non-ASCII hostnames
  /// have been converted to punycode.
  ///
  /// Specifying a `nullptr` for the uri is equivalent to an empty string which
  /// matches no URIs.
  ///
  /// For more information about resource context filters, navigate to
  /// [COREWEBVIEW2_WEB_RESOURCE_CONTEXT].
  ///

>   /// Because service worker runs separately from any one HTML document its WebResourceRequested

Reviewer: also shared workers, correct? 
```suggestion
  /// Because service workers and shared workers run separately from any one HTML document, its WebResourceRequested
```

Author: Ok, you are right - it is how shared workers behave now also - we receive events in both webview2s if they are subscribed even if it is initiated only from the one of them. But I'm wondering if I need to fix this and have shared worker event send to only webview which started it - similar to how it works for main page and iframes? Or it is correct and expected? CC: @LiangTheDev

Reviewer: Shared workers are shared amongst different browsing contexts and so can be used by different CoreWebView2s similarly to service workers so makes sense to me that they raise events on multiple CoreWebView2s (or at least that the behavior matches service workers)

  /// Removes a matching WebResource filter that was previously added for the
  /// `WebResourceRequested` event.  If the same filter was added multiple
  /// times, then it must be removed as many times as it was added for the
  /// removal to be effective.  Returns `E_INVALIDARG` for a filter that was
  /// never added.
  HRESULT RemoveWebResourceRequestedFilterWithRequestSource(
      [in] LPCWSTR const uri,
      [in] COREWEBVIEW2_WEB_RESOURCE_CONTEXT const resourceContext,
      [in] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE const requestSource);
 }

/// Event args for the `WebResourceRequested` event.
[uuid(572f38f9-c317-4b6c-8dd2-c8b894779bbb), object, pointer_default(unique)]
interface ICoreWebView2WebResourceRequestedEventArgs : IUnknown {
  /// The web resource request source.
  [propget] HRESULT RequestSource(
  [out, retval] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE* requestSource);
}
```

## .NET and WinRT
```c# (but really MIDL3)
namespace Microsoft.Web.WebView2.Core
{
    [flags]
    enum CoreWebView2WebResourceRequestSource
    {
        Document = 0x00000001,
        SharedWorker = 0x00000002,
        ServiceWorker = 0x00000004,
        All = 0xFFFFFFFF,
    };

>     [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2_Manual")]

Reviewer: This block should be within the CoreWebView2 runtimeclass

    {
        void AddWebResourceRequestedFilter(
                String uri,

>                 CoreWebView2WebResourceContext ResourceContext,

Reviewer: Lower camel case for parameter names pls
```suggestion
                CoreWebView2WebResourceContext resourceContext,
```

                CoreWebView2WebResourceRequestSource RequestSource);

        void RemoveWebResourceRequestedFilter(
                String uri,
                CoreWebView2WebResourceContext ResourceContext,
                CoreWebView2WebResourceRequestSource RequestSource);
    }

>     [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2StagingWebResourceRequestedEventArgs")]

Reviewer: This block should be within the CoreWebView2WebResourceRequestedEventArgs runtimeclass

                        Microsoft::WRL::ComPtr<ICoreWebView2_2> webview2;
                        Microsoft::WRL::ComPtr<ICoreWebView2Environment> environment;
                        CHECK_FAILURE(m_webviewEventSource->QueryInterface(IID_PPV_ARGS(&webview2)));
                        CHECK_FAILURE(webview2->get_Environment(&environment));
                        CHECK_FAILURE(SHCreateStreamOnFileEx(
                            file_path, STGM_READ, FILE_ATTRIBUTE_NORMAL,
                            FALSE, nullptr, &response_stream));
                        Microsoft::WRL::ComPtr<ICoreWebView2WebResourceResponse> response;
                        CHECK_FAILURE(environment->CreateWebResourceResponse(
                            response_stream.Get(), 200, L"OK", L"Content-Type: image/jpeg",
                            &response));
                        args->put_Response(response.Get());
                    }
                }
                return S_OK;
            })
            .Get(),
        &m_webResourceRequestedToken);
```

# API Details
## IDL
```c#
/// Specifies the source of `WebResourceRequested` event.
[v1_enum]

> typedef enum COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE {

Reviewer: API guidance says to use plural for flags enums (like this). Also this type isn't the source but the kind of source. So we should add _KINDS to the enum type name. (and the enum value names similarly get updated)

```suggestion
typedef enum COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_KINDS {
```

  /// | `https://&#x2764;.example/` | `https://xn--qei.example/` | No | Non-ASCII hostnames are normalized to punycode before wildcard comparison |
  HRESULT AddWebResourceRequestedFilterWithRequestSource(
    [in] LPCWSTR const uri,
    [in] COREWEBVIEW2_WEB_RESOURCE_CONTEXT const resourceContext,
    [in] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE const requestSource);

  /// Removes a matching WebResource filter that was previously added for the
  /// `WebResourceRequested` event.  If the same filter was added multiple
  /// times, then it must be removed as many times as it was added for the
  /// removal to be effective.  Returns `E_INVALIDARG` for a filter that was
  /// never added.
  HRESULT RemoveWebResourceRequestedFilterWithRequestSource(
      [in] LPCWSTR const uri,
      [in] COREWEBVIEW2_WEB_RESOURCE_CONTEXT const resourceContext,
      [in] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE const requestSource);
 }

/// Event args for the `WebResourceRequested` event.
[uuid(572f38f9-c317-4b6c-8dd2-c8b894779bbb), object, pointer_default(unique)]
interface ICoreWebView2WebResourceRequestedEventArgs : IUnknown {
  /// The web resource request source.

>   [propget] HRESULT RequestSource(

Author: @david-risney Do we need to rename this arg to reflect "KINDS" in enum -  `RequestSourceKind`, `SourceKind`?

  [out, retval] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE* requestSource);
}
```

## .NET and WinRT
```c# (but really MIDL3)
namespace Microsoft.Web.WebView2.Core
{
    [flags]
    enum CoreWebView2WebResourceRequestSource
    {
        Document = 0x00000001,
        SharedWorker = 0x00000002,
        ServiceWorker = 0x00000004,
        All = 0xFFFFFFFF,
    };

    runtimeclass CoreWebView2
    {

>         [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2_Manual")]

Reviewer: Nice to add comment noting that we're leaving out unmodified sections of the runtimeclass.
```suggestion
        // ...
        [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2_Manual")]
```

        {
            void AddWebResourceRequestedFilter(
                    String uri,
                    CoreWebView2WebResourceContext resourceContext,
                    CoreWebView2WebResourceRequestSource requestSource);

            void RemoveWebResourceRequestedFilter(
                    String uri,
                    CoreWebView2WebResourceContext resourceContext,
                    CoreWebView2WebResourceRequestSource requestSource);
        }
    }

    runtimeclass CoreWebView2WebResourceRequestedEventArgs
    {

>         [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2StagingWebResourceRequestedEventArgs")]

Reviewer: ```suggestion
        // ...
        [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2StagingWebResourceRequestedEventArgs")]
```

WebResourceRequested events for workers
===

# Background

> Currently [WebResourceRequested events](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#add_webresourcerequested) are raised only for main page HTTP requests. We are extending [AddWebResourceRequestedFilter](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#addwebresourcerequestedfilter) API to allow you additionally to subscribe to events raised from Service or Shared workers and fixing oop iframes events as part of the main page.

Reviewer: Replace 'main page' with 'top level document' here and at least one other place in this document.

```suggestion
Currently [WebResourceRequested events](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#add_webresourcerequested) are raised only for top level document HTTP requests. We are extending [AddWebResourceRequestedFilter](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#addwebresourcerequestedfilter) API to allow you additionally to subscribe to events raised from Service or Shared workers and fixing oop iframes events as part of the main page.
```

> Currently [WebResourceRequested events](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#add_webresourcerequested) are raised only for main page HTTP requests. We are extending [AddWebResourceRequestedFilter](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#addwebresourcerequestedfilter) API to allow you additionally to subscribe to events raised from Service or Shared workers and fixing oop iframes events as part of the main page.

Reviewer: ```suggestion
Currently [WebResourceRequested events](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#add_webresourcerequested) are raised only for main page HTTP requests. We are extending [AddWebResourceRequestedFilter](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/iwebview2webview5?view=webview2-0.8.355#addwebresourcerequestedfilter) API to allow you additionally to subscribe to events raised from Service or Shared workers and fixing an issue where out of process iframes did not raise WebResourceRequested events as part of the main page.
```

  /// | `https://&#x2764;.example/` | `https://xn--qei.example/` | No | Non-ASCII hostnames are normalized to punycode before wildcard comparison |
  HRESULT AddWebResourceRequestedFilterWithRequestSourceKinds(
    [in] LPCWSTR const uri,
    [in] COREWEBVIEW2_WEB_RESOURCE_CONTEXT const resourceContext,
    [in] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_KINDS const requestSourceKinds);

  /// Removes a matching WebResource filter that was previously added for the
  /// `WebResourceRequested` event.  If the same filter was added multiple
  /// times, then it must be removed as many times as it was added for the
  /// removal to be effective.  Returns `E_INVALIDARG` for a filter that was
  /// never added.
  HRESULT RemoveWebResourceRequestedFilterWithRequestSourceKinds(
      [in] LPCWSTR const uri,
      [in] COREWEBVIEW2_WEB_RESOURCE_CONTEXT const resourceContext,
      [in] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_KINDS const requestSourceKinds);
 }

/// Event args for the `WebResourceRequested` event.
[uuid(572f38f9-c317-4b6c-8dd2-c8b894779bbb), object, pointer_default(unique)]

> interface ICoreWebView2WebResourceRequestedEventArgs : IUnknown {

Reviewer: This is an extension of an existing interface right?
```suggestion
interface ICoreWebView2WebResourceRequestedEventArgs2 : ICoreWebView2WebResourceRequestedEventArgs {
```

  /// The web resource request source.
  [propget] HRESULT RequestSourceKinds(
  [out, retval] COREWEBVIEW2_WEB_RESOURCE_REQUEST_SOURCE_KINDS* requestSourceKinds);
}
```

## .NET and WinRT
```c# (but really MIDL3)
namespace Microsoft.Web.WebView2.Core
{
    [flags]
    enum CoreWebView2WebResourceRequestSourceKinds
    {
        Document = 0x00000001,
        SharedWorker = 0x00000002,
        ServiceWorker = 0x00000004,
        All = 0xFFFFFFFF,
    };

    runtimeclass CoreWebView2
    {
        // ...
        [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2_Manual")]
        {
            void AddWebResourceRequestedFilter(
                    String uri,
                    CoreWebView2WebResourceContext resourceContext,
                    CoreWebView2WebResourceRequestSourceKinds requestSourceKinds);

            void RemoveWebResourceRequestedFilter(
                    String uri,
                    CoreWebView2WebResourceContext resourceContext,
                    CoreWebView2WebResourceRequestSourceKinds requestSourceKinds);
        }
    }

    runtimeclass CoreWebView2WebResourceRequestedEventArgs
    {
        // ...

>         [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2StagingWebResourceRequestedEventArgs")]

Reviewer: ```suggestion
        [interface_name("Microsoft.Web.WebView2.Core.ICoreWebView2WebResourceRequestedEventArgs2")]
```
